File: hostel_management_system_backend/Services/LocalImageStorageService.cs
using hostel_management_system_backend.Exceptions;
using Microsoft.AspNetCore.Http;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Formats;
using SixLabors.ImageSharp.Formats.Webp;
using SixLabors.ImageSharp.Processing;

public sealed class LocalImageStorageService : IImageStorageService
{
    private const long MaxImageSizeBytes = 5 * 1024 * 1024;
    private const string StoredContentType = "image/webp";

    private static readonly HashSet<string> AllowedContentTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "image/jpeg",
        "image/png",
        "image/webp"
    };

    private readonly IWebHostEnvironment _environment;
    private readonly IConfiguration _configuration;

    public LocalImageStorageService(IWebHostEnvironment environment, IConfiguration configuration)
    {
        _environment = environment;
        _configuration = configuration;
    }

    public async Task<StoredImageResult> UploadImageAsync(IFormFile file, Guid hostelId, CancellationToken cancellationToken)
    {
        ValidateImage(file);

        var webRoot = string.IsNullOrWhiteSpace(_environment.WebRootPath)
            ? Path.Combine(_environment.ContentRootPath, "wwwroot")
            : _environment.WebRootPath;

        var hostelFolder = Path.Combine(webRoot, "uploads", "hostels", hostelId.ToString());
        var thumbnailFolder = Path.Combine(hostelFolder, "thumbnail");
        var cardFolder = Path.Combine(hostelFolder, "card");
        var fullFolder = Path.Combine(hostelFolder, "full");
        Directory.CreateDirectory(thumbnailFolder);
        Directory.CreateDirectory(cardFolder);
        Directory.CreateDirectory(fullFolder);

        var baseFileName = Guid.NewGuid().ToString("N");
        var storedFileName = $"{baseFileName}.webp";

        var thumbnailRelative = $"/uploads/hostels/{hostelId}/thumbnail/{storedFileName}";
        var cardRelative = $"/uploads/hostels/{hostelId}/card/{storedFileName}";
        var fullRelative = $"/uploads/hostels/{hostelId}/full/{storedFileName}";

        var thumbnailPath = Path.Combine(thumbnailFolder, storedFileName);
        var cardPath = Path.Combine(cardFolder, storedFileName);
        var fullPath = Path.Combine(fullFolder, storedFileName);

        await using var uploadStream = file.OpenReadStream();

        Image image;
        try
        {
            image = await Image.LoadAsync(uploadStream, cancellationToken);
        }
        catch (UnknownImageFormatException)
        {
            throw new BadRequestException("Invalid image type. Allowed types are jpeg, png, webp.");
        }

        using (image)
        {
            var encoder = new WebpEncoder { Quality = 80 };
            await SaveResizedAsync(image, 300, thumbnailPath, encoder, cancellationToken);
            await SaveResizedAsync(image, 600, cardPath, encoder, cancellationToken);
            await SaveResizedAsync(image, 1200, fullPath, encoder, cancellationToken);
        }

        var fullImageFileInfo = new FileInfo(fullPath);
        var imageUrl = BuildPublicUrl(fullRelative);

        return new StoredImageResult(
            imageUrl,
            StoredContentType,
            fullImageFileInfo.Exists ? fullImageFileInfo.Length : 0,
            storedFileName
        );
    }

    public Task<bool> DeleteImageAsync(string imageUrl, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(imageUrl))
        {
            return Task.FromResult(false);
        }

        var imagePath = ExtractPath(imageUrl);
        if (string.IsNullOrWhiteSpace(imagePath) || !imagePath.StartsWith('/'))
        {
            return Task.FromResult(false);
        }

        var webRoot = string.IsNullOrWhiteSpace(_environment.WebRootPath)
            ? Path.Combine(_environment.ContentRootPath, "wwwroot")
            : _environment.WebRootPath;

        var deleted = false;
        foreach (var relativeVariantPath in GetVariantRelativePaths(imagePath))
        {
            var relativePath = relativeVariantPath.TrimStart('/').Replace('/', Path.DirectorySeparatorChar);
            var fullPath = Path.Combine(webRoot, relativePath);
            if (!File.Exists(fullPath))
            {
                continue;
            }

            File.Delete(fullPath);
            deleted = true;
        }

        return Task.FromResult(deleted);
    }

    private static void ValidateImage(IFormFile file)
    {
        if (file.Length <= 0)
        {
            throw new BadRequestException("Image file is required.");
        }

        if (file.Length > MaxImageSizeBytes)
        {
            throw new BadRequestException("Image too large. Maximum allowed size is 5MB.");
        }

        if (string.IsNullOrWhiteSpace(file.ContentType) || !AllowedContentTypes.Contains(file.ContentType))
        {
            throw new BadRequestException("Invalid image type. Allowed types are jpeg, png, webp.");
        }
    }

    private static async Task SaveResizedAsync(
        Image original,
        int width,
        string outputPath,
        IImageEncoder encoder,
        CancellationToken cancellationToken)
    {
        using var resized = original.Clone(ctx =>
            ctx.Resize(new ResizeOptions
            {
                Size = new Size(width, 0),
                Mode = ResizeMode.Max
            }));

        await using var outputStream = new FileStream(outputPath, FileMode.Create, FileAccess.Write, FileShare.None);
        await resized.SaveAsync(outputStream, encoder, cancellationToken);
    }

    private string BuildPublicUrl(string relativePath)
    {
        var cdnBaseUrl = _configuration["ImageStorage:CdnBaseUrl"];
        if (string.IsNullOrWhiteSpace(cdnBaseUrl))
        {
            return relativePath;
        }

        return $"{cdnBaseUrl.TrimEnd('/')}{relativePath}";
    }

    private static string? ExtractPath(string imageUrl)
    {
        if (Uri.TryCreate(imageUrl, UriKind.Absolute, out var absoluteUri))
        {
            return absoluteUri.AbsolutePath;
        }

        return imageUrl;
    }

    private static IEnumerable<string> GetVariantRelativePaths(string imagePath)
    {
        if (imagePath.Contains("/full/", StringComparison.OrdinalIgnoreCase))
        {
            yield return imagePath;
            yield return imagePath.Replace("/full/", "/card/", StringComparison.OrdinalIgnoreCase);
            yield return imagePath.Replace("/full/", "/thumbnail/", StringComparison.OrdinalIgnoreCase);
            yield break;
        }

        if (imagePath.Contains("/card/", StringComparison.OrdinalIgnoreCase))
        {
            yield return imagePath;
            yield return imagePath.Replace("/card/", "/full/", StringComparison.OrdinalIgnoreCase);
            yield return imagePath.Replace("/card/", "/thumbnail/", StringComparison.OrdinalIgnoreCase);
            yield break;
        }

        if (imagePath.Contains("/thumbnail/", StringComparison.OrdinalIgnoreCase))
        {
            yield return imagePath;
            yield return imagePath.Replace("/thumbnail/", "/full/", StringComparison.OrdinalIgnoreCase);
            yield return imagePath.Replace("/thumbnail/", "/card/", StringComparison.OrdinalIgnoreCase);
            yield break;
        }

        yield return imagePath;
    }
}

File: src/hostels/images/local-image-storage.service.ts
import { mkdir, stat, unlink } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { randomUUID } from 'node:crypto';
import { join } from 'node:path';

import { Injectable } from '@nestjs/common';
import sharp from 'sharp';

import { AppBadRequestException } from '../../common/exceptions/app-exception';
import { AppConfigService } from '../../config/app-config.service';

export interface StoredImageResult {
  imageUrl: string;
  contentType: string;
  fileSize: number;
  storedFileName: string;
}

@Injectable()
export class LocalImageStorageService {
  private static readonly MAX_IMAGE_SIZE_BYTES = 5 * 1024 * 1024;
  private static readonly ALLOWED_CONTENT_TYPES = new Set(['image/jpeg', 'image/png', 'image/webp']);

  constructor(private readonly configService: AppConfigService) {}

  async uploadImage(file: Express.Multer.File, hostelId: string): Promise<StoredImageResult> {
    this.validateImage(file);

    const hostelFolder = join(process.cwd(), 'wwwroot', 'uploads', 'hostels', hostelId);
    const thumbnailFolder = join(hostelFolder, 'thumbnail');
    const cardFolder = join(hostelFolder, 'card');
    const fullFolder = join(hostelFolder, 'full');

    await Promise.all([
      mkdir(thumbnailFolder, { recursive: true }),
      mkdir(cardFolder, { recursive: true }),
      mkdir(fullFolder, { recursive: true }),
    ]);

    const storedFileName = `${randomUUID().replace(/-/g, '')}.webp`;
    const thumbnailPath = join(thumbnailFolder, storedFileName);
    const cardPath = join(cardFolder, storedFileName);
    const fullPath = join(fullFolder, storedFileName);

    await Promise.all([
      this.saveResized(file.buffer, 300, thumbnailPath),
      this.saveResized(file.buffer, 600, cardPath),
      this.saveResized(file.buffer, 1200, fullPath),
    ]);

    const fullStat = await stat(fullPath);
    const relativePath = `/uploads/hostels/${hostelId}/full/${storedFileName}`;

    return {
      imageUrl: this.buildPublicUrl(relativePath),
      contentType: 'image/webp',
      fileSize: fullStat.size,
      storedFileName,
    };
  }

  async deleteImage(imageUrl: string): Promise<boolean> {
    if (!imageUrl?.trim()) {
      return false;
    }

    const imagePath = this.extractPath(imageUrl);
    if (!imagePath || !imagePath.startsWith('/')) {
      return false;
    }

    const webRoot = join(process.cwd(), 'wwwroot');
    let deleted = false;

    for (const relativePath of this.getVariantRelativePaths(imagePath)) {
      const fullPath = join(webRoot, relativePath.replace(/^\//, '').replaceAll('/', '\\'));
      if (!existsSync(fullPath)) {
        continue;
      }

      await unlink(fullPath);
      deleted = true;
    }

    return deleted;
  }

  private validateImage(file: Express.Multer.File) {
    if (!file || file.size <= 0) {
      throw new AppBadRequestException('Image file is required.');
    }

    if (file.size > LocalImageStorageService.MAX_IMAGE_SIZE_BYTES) {
      throw new AppBadRequestException('Image too large. Maximum allowed size is 5MB.');
    }

    if (!LocalImageStorageService.ALLOWED_CONTENT_TYPES.has(file.mimetype)) {
      throw new AppBadRequestException('Invalid image type. Allowed types are jpeg, png, webp.');
    }
  }

  private async saveResized(buffer: Buffer, width: number, outputPath: string) {
    await sharp(buffer)
      .rotate()
      .resize({
        width,
        fit: 'inside',
        withoutEnlargement: true,
      })
      .webp({ quality: 80 })
      .toFile(outputPath);
  }

  private buildPublicUrl(relativePath: string) {
    return this.configService.cdnBaseUrl
      ? `${this.configService.cdnBaseUrl.replace(/\/$/, '')}${relativePath}`
      : relativePath;
  }

  private extractPath(imageUrl: string) {
    try {
      return new URL(imageUrl).pathname;
    } catch {
      return imageUrl;
    }
  }

  private *getVariantRelativePaths(imagePath: string) {
    if (imagePath.includes('/full/')) {
      yield imagePath;
      yield imagePath.replace('/full/', '/card/');
      yield imagePath.replace('/full/', '/thumbnail/');
      return;
    }

    if (imagePath.includes('/card/')) {
      yield imagePath;
      yield imagePath.replace('/card/', '/full/');
      yield imagePath.replace('/card/', '/thumbnail/');
      return;
    }

    if (imagePath.includes('/thumbnail/')) {
      yield imagePath;
      yield imagePath.replace('/thumbnail/', '/full/');
      yield imagePath.replace('/thumbnail/', '/card/');
      return;
    }

    yield imagePath;
  }
}
